src/codegen/build_module.cc

```diff
@@ -81,33 +87,65 @@ Target Target::create(const std::string& target_str) {
   return result;
 }
 
+/*! \brief Entry to hold the Target context stack. */
+struct TVMTargetThreadLocalEntry {
+  /*! \brief The current target context */
+  std::stack<tvm::Target> context_stack;
+
+  TVMTargetThreadLocalEntry() {
+  }
+};
+
+/*! \brief Thread local store to hold the Target context stack. */
+typedef dmlc::ThreadLocalStore<TVMTargetThreadLocalEntry> TVMTargetThreadLocalStore;
+
+void Target::EnterTargetScope(const tvm::Target& target) {
+  TVMTargetThreadLocalEntry *entry = TVMTargetThreadLocalStore::Get();
+  entry->context_stack.push(target);
+}
+
+void Target::ExitTargetScope() {
+  TVMTargetThreadLocalEntry *entry = TVMTargetThreadLocalStore::Get();
+  entry->context_stack.pop();
+}
+
+tvm::Target* Target::current_target(bool allow_null) {
```

Comment:
This is because it needs to be able to return nullptr if the stack is empty and allow_null is true. This is used preserve the behavior of python current_target() when the allow_none argument is true.